src/framework/constraints.ml

```diff
@@ -1694,6 +1694,162 @@ struct
   let event ctx e octx = S.event (conv ctx) e (conv octx)
 end
 
+
+(** Add cycle detection in the context-sensitive dynamic function call graph to an analysis *)
+module RecursionTermLifter (S: Spec)
+  : Spec with module D = S.D
+          and module C = S.C
+=
+(* two global invariants:
+   - V -> G
+     Needed to store the previously built global invariants
+   - fundec -> Map (S.C) (Set (fundec * S.C))
+     The second global invariant maps from the callee fundec to a map, containing the callee context and the caller fundec and context.
+     This structure therefore stores the context-sensitive call graph.
+     For example:
+      let the function f in context c call function g in context c'.
+      In the global invariant structure it would be stored like this: g -> {c' -> {(f, c)}}
```

Comment:
Why not use `fundec * S.C -> Set (fundec * S.C)` as the invariant?

Reply:
Fixed!